CAIPs/caip-x.md

```diff
@@ -0,0 +1,339 @@
+---
+caip: XX
+title: JSON-RPC Provider Lifecycle Methods for Session Management
+author: [Alex Donesky] (@adonesky1)
+discussions-to: TBD
+status: Draft
+type: Standard
+created: 2024-06-07
+requires: 25, 217
+---
+
+## Simple Summary
+
+CAIP-XX introduces new "lifecycle" methods and a new event for managing permissions within an existing CAIP-25 session, allowing for addition, revocation, and retrieval of permissions. These methods provide an alternative to session management via `sessionId`s, allowing `sessionId`s to be optional for CAIP-25.
+
+## Abstract
+
+This proposal aims to extend the CAIP-25 standard by defining new JSON-RPC methods for managing the lifecycle of permissions within a session. These methods allow dapps and wallets to dynamically adjust permissions, providing more granular control and better user experience. Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.
+
+## Motivation
+
+The motivation behind this proposal is to enhance the flexibility of CAIP-25 by enabling management of session permissions without sessionIds which don't map well to extension based wallet's dapp connections and could therefore add constraints and burdens to existing flows. The proposed methods provide an intuitive way to add, revoke, and retrieve permissions within an existing session, simplifying the management of session lifecycles.
+
+### Use Case Scenarios
+
+1. **Dapp Initiated Adding Permissions To an Existing Session:**
+
+   - **Current Method:** Call `provider_authorize` again with an existing session identifier and new scopes/methods to add. This is actually somewhat ambiguous in CAIP-25 where it’s unclear if incremental permissions adds should include the full object or just scopes to add: "This object gets updated, extended, closed, etc. by successive calls and notifications, each tagged by this identifier."
+   - **Proposed Method:** Use `provider_augment` to add new permissions to the existing session. Alternatively, this could be achieved with more specific language indicating that a subsequent `provider_authorize` call with new scopes/permissions could simply add to an existing session. If we pursue this latter approach one flow that will need clarity is what happens if a subsequent `provider_authorize` request includes `requiredScopes` which the respondent does not authorize. In this case I would expect that this does not revoke the existing session but rather only the new request is rejected.
+
+2. **Wallet Initiated Adding Permissions To an Existing Session:**
```

Comment:
What's the use case here?

Reply:
This kindof thing:
https://github.com/ChainAgnostic/CAIPs/assets/34557516/bd22d855-ac36-40b4-ad4f-b186d4c060af

Reply:
i don't understand

Reply:
In the video I add a permission/authorization for the site (in this case github) to use Account 3 all in the wallet (without any RPC Request Response cycle) and the site in this case (though it is not depicted) received an `accountsChanged` [event](https://docs.metamask.io/wallet/reference/provider-api/#accountschanged). This scenario `**Wallet Initiated Adding Permissions To an Existing Session:**` and the use of the `providerAuthorizationUpdated` would provide the equivalent functionality

Reply:
So, just to be clear, is the idea here that the existing `accountsChanged` notification that some dapps already support, which can be authorized on eip155 `scopeObjects`, would trigger a change in the CAIP-25 config (i.e. add a member to the `chains` array in a `scopeObject` scoped to all of `eip155`), or that a net-new notification (mandatory to support for all CAIP-25 wallets on any chain) would replace it, or both? I can see pros and cons to all three, just want everything as explicit as possible for the sake of minimizing and tracking all possible breakage and corner cases.

This is currently possible connecting, say, MME over WC, but not AFAIK load-bearing in any dapps.

Reply:
So the purpose of this notification is to provide the wallet (or respondent) to modify (or signal the modification of) the session to the dapp (or caller). `accountsChanged` is limited to signaling to dApps that they have been granted permissions to new accounts. This event could be used to indicate any changes made by the user of the wallet to the permissions/authorizations that make up the session. The closest that CAIP-25 gets to providing this flow is:

> "If a respondent (e.g. a wallet) needs to initiate a new session, whether due to user input, security policy, or session expiry reasons, it can simply generate a new session identifier to signal this notification to the calling provider; if a caller needs to initiate a new session, it can do so by sending a new request without a sessionIdentifier. In such cases, a respondent (e.g. wallet) may choose to explicitly close all sessions upon generation of a new one from the same origin or identity, or leave it to time-out; maintaining concurrent sessions is discouraged (see Security Considerations)."

But in this case any modifications made on the wallet side would require tearing down the existing session and generating a new one - and it relies on sessionIds which we're looking to make optional. This doesn't seem very ergonomic or efficient. Let me know if I'm missing something.